lib/core/cmap/connection.js

```
@@ -0,0 +1,211 @@
+    this[kMessageStream].on('message', messageHandler(this));
+    this[kStream] = stream;
+    stream.on('error', () => {
+      /* ignore errors, listen to `close` instead */
+    });
+
+    stream.on('close', () => {
+      this[kQueue].forEach(op => op.callback(new MongoError('Connection closed')));
+      this[kQueue].clear();
+
+      this.emit('close');
+    });
+
+    // hook the message stream up to the passed in stream
+    stream.pipe(this[kMessageStream]);
+    this[kMessageStream].pipe(stream);
+  }
+
+  // the `connect` method stores the result of the handshake ismaster on the connection
+  set ismaster(response) {
+    this.description = response;
+  }
+
+  destroy(options, callback) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+
+    options = Object.assign({ force: false }, options);
+    if (this[kStream] == null || this.destroyed) {
+      this.destroyed = true;
+      return;
+    }
+
+    if (options.force) {
+      this[kStream].destroy();
+      this.destroyed = true;
+      if (typeof callback === 'function') {
+        callback(null, null);
+      }
+
+      return;
+    }
+
+    this[kStream].end(err => {
+      this.destroyed = true;
+      if (typeof callback === 'function') {
+        callback(err, null);
+      }
+    });
+  }
+
+  command(ns, cmd, options, callback) {
+    // NOTE: The wire protocol methods will eventually be migrated to this class, but for now
+    //       we need to pretend we _are_ a server.
+    const server = {
+      description: this.description,
+      s: {
+        bson: this.bson,
+        pool: { write: write.bind(this) }
+      }
+    };
+
+    wp.command(server, ns, cmd, options, callback);
+  }
+}
+
+function messageHandler(conn) {
+  return function(message) {
+    // always emit the message, in case we are streaming
+    conn.emit('message', message);
+    if (!conn[kQueue].has(message.responseTo)) {
+      return;
```

Seems weird to silently drop in this case.

---

Also what about a `moreToCome`?